# Pandas stands for Python Data Analysis Library

# What is Library ? 

# Library is actually a set of predefined functions developed to achieve a specific class of operations

# Now, using Pandas Library, you actually work on data and get valuable descriptive as well inferential statistics from your data

# The types of data which we will be dealing with in PGDIDS&ML is going to be four types:

# Tabular Data (In the form of tables, excel files, csv file,.......etc)

# Textual Data (In the form of documents, word files, pdfs, notepad files...etc)

# Images Data (In the form of jpeg, png files)

# Videos (In the form of mp4 or whatever)

# Pandas library is primarily used to work on tabular data

# CSV stands for Comma Seperated Value Format. It is just like your excel format. 

# If we want to do any kind of analysis on csv data then we need to first of all bring that data into python environment and for that we have to read the data, that means we have to first fetch everything in tabular data format file. Now, let's see how can we do that: 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("mushrooms.csv")

In [3]:
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b           s         y       t    a               f   
13       p         x           y         w       t    p               f   
14       e         x           f         n       f    n               f   
15       e         s           f         g       f    n               f   
16       e         f           f         w       f    n               f   
17       p         x           s         n       t    p               f   
18       p         x           y         w       t    p               f   
19       p         x           s         n       t    p               f   
20       e         b           s         y       t    a               f   
21       p         x           y         n       t    p               f   
22       e         b           y         y       t    l               f   
23       e         b           y         w       t    a               f   
24       e         b           s         w       t    l               f   
25       p         f           s         w       t    p               f   
26       e         x           y         y       t    a               f   
27       e         x           y         w       t    l               f   
28       e         f           f         n       f    n               f   
29       e         x           s         y       t    a               f   
...    ...       ...         ...       ...     ...  ...             ...   
8094     e         b           s         g       f    n               f   
8095     p         x           y         c       f    m               f   
8096     e         k           f         w       f    n               f   
8097     p         k           y         n       f    s               f   
8098     p         k           s         e       f    y               f   
8099     e         k           f         w       f    n               f   
8100     e         f           s         n       f    n               a   
8101     p         k           s         e       f    s               f   
8102     e         x           s         n       f    n               a   
8103     e         k           s         n       f    n               a   
8104     e         k           s         n       f    n               a   
8105     e         k           s         n       f    n               a   
8106     e         k           s         n       f    n               a   
8107     e         x           s         n       f    n               a   
8108     p         k           y         e       f    y               f   
8109     e         b           s         w       f    n               f   
8110     e         x           s         n       f    n               a   
8111     e         k           s         w       f    n               f   
8112     e         k           s         n       f    n               a   
8113     p         k           y         e       f    y               f   
8114     p         f           y         c       f    m               a   
8115     e         x          

# The way your data is getting displayed above is a different way of storing the data apart from lists, dictionaries, tuples and sets. This way of storing the data is Pandas DataFrame. A Pandas DataFrame is another kind of Data Structure. 

In [4]:
df.iloc[7:13]

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
7      e         b           y         w       t    l               f   
8      p         x           y         w       t    p               f   
9      e         b           s         y       t    a               f   
10     e         x           y         y       t    l               f   
11     e         x           y         y       t    a               f   
12     e         b           s         y       t    a               f   

   gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
7             c         b          n  ...                        s   
8             c         n          p  ...                        s   
9             c         b          g  ...                        s   
10            c         b          g  ...                        s   
11            c         b          n  ...                        s   
12            c         b          w  ...                        s   

   stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
7                       w                      w         p          w   
8                       w                      w         p          w   
9                       w                      w         p          w   
10                      w                      w         p          w   
11                      w                      w         p          w   
12                      w                      w         p          w   

   ring-number ring-type spore-print-color population habitat  
7            o         p                 n          s       m  
8            o         p                 k          v       g  
9            o         p                 k          s       m  
10           o         p                 n          n       g  
11           o         p                 k          s       m  
12           o         p                 n          s       g  

[6 rows x 23 columns]

# In the above manner, you can access rows of the Pandas DataFrame using iloc function

# Now let's see that how can we access columns of a Pandas DataFrame

In [5]:
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b           s         y       t    a               f   
13       p         x           y         w       t    p               f   
14       e         x           f         n       f    n               f   
15       e         s           f         g       f    n               f   
16       e         f           f         w       f    n               f   
17       p         x           s         n       t    p               f   
18       p         x           y         w       t    p               f   
19       p         x           s         n       t    p               f   
20       e         b           s         y       t    a               f   
21       p         x           y         n       t    p               f   
22       e         b           y         y       t    l               f   
23       e         b           y         w       t    a               f   
24       e         b           s         w       t    l               f   
25       p         f           s         w       t    p               f   
26       e         x           y         y       t    a               f   
27       e         x           y         w       t    l               f   
28       e         f           f         n       f    n               f   
29       e         x           s         y       t    a               f   
...    ...       ...         ...       ...     ...  ...             ...   
8094     e         b           s         g       f    n               f   
8095     p         x           y         c       f    m               f   
8096     e         k           f         w       f    n               f   
8097     p         k           y         n       f    s               f   
8098     p         k           s         e       f    y               f   
8099     e         k           f         w       f    n               f   
8100     e         f           s         n       f    n               a   
8101     p         k           s         e       f    s               f   
8102     e         x           s         n       f    n               a   
8103     e         k           s         n       f    n               a   
8104     e         k           s         n       f    n               a   
8105     e         k           s         n       f    n               a   
8106     e         k           s         n       f    n               a   
8107     e         x           s         n       f    n               a   
8108     p         k           y         e       f    y               f   
8109     e         b           s         w       f    n               f   
8110     e         x           s         n       f    n               a   
8111     e         k           s         w       f    n               f   
8112     e         k           s         n       f    n               a   
8113     p         k           y         e       f    y               f   
8114     p         f           y         c       f    m               a   
8115     e         x          

In [6]:
df_columns = list(df.columns)

In [7]:
df_columns

['class',
 'cap-shape',
 'cap-surface',
 'cap-color',
 'bruises',
 'odor',
 'gill-attachment',
 'gill-spacing',
 'gill-size',
 'gill-color',
 'stalk-shape',
 'stalk-root',
 'stalk-surface-above-ring',
 'stalk-surface-below-ring',
 'stalk-color-above-ring',
 'stalk-color-below-ring',
 'veil-type',
 'veil-color',
 'ring-number',
 'ring-type',
 'spore-print-color',
 'population',
 'habitat']

In [8]:
df['cap-surface']

0       s
1       s
2       s
3       y
4       s
5       y
6       s
7       y
8       y
9       s
10      y
11      y
12      s
13      y
14      f
15      f
16      f
17      s
18      y
19      s
20      s
21      y
22      y
23      y
24      s
25      s
26      y
27      y
28      f
29      s
       ..
8094    s
8095    y
8096    f
8097    y
8098    s
8099    f
8100    s
8101    s
8102    s
8103    s
8104    s
8105    s
8106    s
8107    s
8108    y
8109    s
8110    s
8111    s
8112    s
8113    y
8114    y
8115    s
8116    y
8117    s
8118    y
8119    s
8120    s
8121    s
8122    y
8123    s
Name: cap-surface, Length: 8124, dtype: object

In [9]:
df.loc[7:12,'cap-surface':'gill-color']

cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size  \
7            y         w       t    l               f            c         b   
8            y         w       t    p               f            c         n   
9            s         y       t    a               f            c         b   
10           y         y       t    l               f            c         b   
11           y         y       t    a               f            c         b   
12           s         y       t    a               f            c         b   

   gill-color  
7           n  
8           p  
9           g  
10          g  
11          n  
12          w

In [10]:
df.iloc[7:13,2:10]

cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size  \
7            y         w       t    l               f            c         b   
8            y         w       t    p               f            c         n   
9            s         y       t    a               f            c         b   
10           y         y       t    l               f            c         b   
11           y         y       t    a               f            c         b   
12           s         y       t    a               f            c         b   

   gill-color  
7           n  
8           p  
9           g  
10          g  
11          n  
12          w

# Suppose, I want to compute the following probability:

\begin{equation}
P(class = 'p')
\end{equation}

# Well, I have a sample of 8124 Mushrooms, but I don't have the population. Therefore, I can calculate the estimate of this probability using the sample. And the estimate is given by the relative frequency. The relative frequency of Poissonous mushrooms is given by: 

\begin{equation}
\hat{P}(class = poissonous) = \frac{Frequency(Mushroom = Poissonous)}{Total Number of Mushrooms (Sample Size)}
\end{equation}

\begin{equation}
\hat{P}(class = poissonous) = \frac{Number of rows corresponsing to poissonous mushrooms in the Pandas DataFrame}{Sample Size (Total Number of rows in Pandas DataFrame)}
\end{equation}

# So, how we are going to compute the Frequency of poissonous mushrooms in our data. For that, we have to seperate the data of all the poissonous mushrooms from our Pandas DataFrame. This seperated data will be in the form of another smaller DataFrame. Then we will be computing the total number of rows in the smaller DataFrame. 

# Now, how we are going to do that ? 

# In order to understand this, we first have to understand the concept of Boolean masks. 

# So, what are Booleans Masks ? 

# Boolean Masks are actually the DataFrame of Boolean values (True and False) where True corresponds to some condition being True and False corresponds to the same condition being False. 

# Now, the next concept after slicing of DataFrame from iloc as well as loc, is going to be of Boolean Masks. 

In [11]:
df_columns[0]

'class'

In [12]:
df['class']

0       p
1       e
2       e
3       p
4       e
5       e
6       e
7       e
8       p
9       e
10      e
11      e
12      e
13      p
14      e
15      e
16      e
17      p
18      p
19      p
20      e
21      p
22      e
23      e
24      e
25      p
26      e
27      e
28      e
29      e
       ..
8094    e
8095    p
8096    e
8097    p
8098    p
8099    e
8100    e
8101    p
8102    e
8103    e
8104    e
8105    e
8106    e
8107    e
8108    p
8109    e
8110    e
8111    e
8112    e
8113    p
8114    p
8115    e
8116    p
8117    p
8118    p
8119    e
8120    e
8121    e
8122    p
8123    e
Name: class, Length: 8124, dtype: object

In [13]:
boolean_mask = df['class'] == 'p'

# Now, as you can that the above DataFrame is a DataFrame of Boolean Values and hence it is called boolean mask. 

# Now, if we pass this boolean mask inside the square brackets of our pandas dataframe:

# df[boolean_mask]

# then, let's see what happens:

In [14]:
poissonous_mushrooms = df[boolean_mask]

# we have to find now out of this data that, how many mushrooms are there which have poissonous category so for that, we have a function associated with the DataFrame and that is called shape. 

In [15]:
poissonous_mushrooms.shape[0]

3916

# So, finally we were able to find the Frequency(Mushroom = Poissonous) = 3916

# So, we have to compute now the total number of mushrooms (sample size) in our data

In [16]:
sample_size = df.shape[0]

In [17]:
sample_size

8124

In [18]:
p_class_equals_poissonous = poissonous_mushrooms.shape[0]/df.shape[0]

In [19]:
p_class_equals_poissonous

0.48202855736090594

# Therefore, 

\begin{equation}
\hat{P}(Class = Poissonous) = 0.48202855736090594
\end{equation}

# Now, let's try to compute another set of Probabilities. Let's say that we want to compute the following probabilities:

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous)
\end{equation}

# The above Probability is actually the Probability of picking up a mushroom whose cap-shape = x as well as cap-surface = y as well as cap-color = w on condition that the mushroom belongs to poissonous category. 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n)
\end{equation}

# The above probability is actually probability of picking up a mushroom whose gill-spacing = c as well as stalk-surface-below-ring = s on a condition that the mushroom belongs to poissonous category as well as it's spore-print-color = n

# So, lets focus on the first type of Probability: 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous)
\end{equation}

# Let's say that we have an event                                                                              A = (cap-shape=x and cap-surface=y and cap-color=w) and similarly, you have another event B = (class = poissonous)

# it means that your event A is actually the event of picking up of random mushroom whose cap-shape=x and cap-surface=y and cap-color=w furthermore, your event B is actually the event of picking up of random mushroom whose category is poissonous. 

# It means that, we can write: 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous) = P(A|B)
\end{equation}

# we know that, according to the formula for conditional Probability: 

\begin{equation}
P(A|B) = \frac{P(A \cap B)}{P(B)}
\end{equation}

# Therefore, 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous) = \frac{P(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap class = poissonous)}{P(class = poissonous)}
\end{equation}

# So, now we have to find two probabilities:

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap class = poissonous)
\end{equation}

# and 

\begin{equation}
P(class = poissonous)
\end{equation}

# We have aleady found out the probability in the denominator, but we have to find the probability in the numerator: 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap class = poissonous)
\end{equation}

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap class = poissonous) = \frac{Frequency(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap clas=poissonous)}{Sample Size (8124)}
\end{equation}

# Which means that, we need to find out:

# Frequency (cap-shape=x and cap-surface=y and cap-color=w and class=poissonous) = Number of rows in pandas dataframe satisfying all the four conditions simultaniously at the same time

# In order to compute this frequency, we need to create the boolean mask of the four conditions simultaniously. 

In [20]:
boolean_mask = (df['cap-shape'] == 'x') & (df['cap-surface'] == 'y') & (df['cap-color'] == 'w') & (df['class'] == 'p')

In [21]:
frequency = df[boolean_mask].shape[0]

In [22]:
frequency

32

# Therefore, 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w \cap class = poissonous) = \frac{32}{Sample Size (8124)}
\end{equation}

In [23]:
joint_probability = df[boolean_mask].shape[0]/df.shape[0]

In [24]:
joint_probability

0.003938946331856229

# Therefore our final conditional probability: 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous)
\end{equation}

# is given by: 

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous) = \frac{joint probability}{p class equals poissonous}
\end{equation}

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous) = \frac{0.003938946331856229}{0.48202855736090594}
\end{equation}

In [25]:
conditional_probability = joint_probability/p_class_equals_poissonous

In [26]:
conditional_probability

0.008171603677221655

\begin{equation}
P(cap-shape=x \cap cap-surface=y \cap cap-color=w | class = poissonous) = 0.008171603677221655
\end{equation}

# So, now we are going to compute the following Probability: 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n)
\end{equation}

# let's say that event A = (gill-spacing=c and stalk-surface-below-ring=s), similarly, event B = (class=poissonous and spore-print-color=n)

# Therefore, we can say that: 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n) = P(A|B)
\end{equation}

# Now, according to conditional probability formula, we know that: 

\begin{equation}
P(A|B) = \frac{P(A \cap B)}{P(B)}
\end{equation}

# Therefore, 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n) = \frac{P(gill-spacing=c \cap stalk-surface-below-ring=s \cap class = poissonous \cap spore-print-color=n)}{P(class = poissonous \cap spore-print-color=n)}
\end{equation}

# Now, we have to compute two probabilities: 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s \cap class = poissonous \cap spore-print-color=n)
\end{equation}

# and

\begin{equation}
P(class = poissonous \cap spore-print-color=n)
\end{equation}

# Let's compute the first (numerator) probability: 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s \cap class = poissonous \cap spore-print-color=n) = \frac{Frequency(gill-spacing=c \cap stalk-surface-below-ring=s \cap class = poissonous \cap spore-print-color=n)}{Sample Size(8124)}
\end{equation}

# We know that now, 

# Frequency(gill-spacing=c and stalk-surface-below-ring=s and class = poissonous and spore-print-color=n) = Number of rows in pandas dataframe satisfying all the four conditions. 

# For getting this frequency, therefore, we have to create the boolean mask

In [27]:
df_columns

['class',
 'cap-shape',
 'cap-surface',
 'cap-color',
 'bruises',
 'odor',
 'gill-attachment',
 'gill-spacing',
 'gill-size',
 'gill-color',
 'stalk-shape',
 'stalk-root',
 'stalk-surface-above-ring',
 'stalk-surface-below-ring',
 'stalk-color-above-ring',
 'stalk-color-below-ring',
 'veil-type',
 'veil-color',
 'ring-number',
 'ring-type',
 'spore-print-color',
 'population',
 'habitat']

In [28]:
df_columns[13]

'stalk-surface-below-ring'

In [29]:
df_columns[20]

'spore-print-color'

In [30]:
boolean_mask = (df['gill-spacing'] == 'c') & (df[df_columns[13]] == 's') & (df[df_columns[0]] == 'p') & (df[df_columns[20]] == 'n')

In [31]:
boolean_mask

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13       True
14      False
15      False
16      False
17      False
18       True
19       True
20      False
21       True
22      False
23      False
24      False
25       True
26      False
27      False
28      False
29      False
        ...  
8094    False
8095    False
8096    False
8097    False
8098    False
8099    False
8100    False
8101    False
8102    False
8103    False
8104    False
8105    False
8106    False
8107    False
8108    False
8109    False
8110    False
8111    False
8112    False
8113    False
8114    False
8115    False
8116    False
8117    False
8118    False
8119    False
8120    False
8121    False
8122    False
8123    False
Length: 8124, dtype: bool

In [32]:
frequency = df[boolean_mask].shape[0]

In [33]:
frequency

176

In [34]:
joint_probability_num = frequency/df.shape[0]

In [35]:
joint_probability_num

0.021664204825209258

# now, similarly I can compute the probability in the denominator: 

\begin{equation}
P(class = poissonous \cap spore-print-color=n) = \frac{Frequency(class=poissonous \cap spore-print-color=n)}{Sample Size(8124)}
\end{equation}

# The frequency above can be computed in the similar manner as before: 

In [36]:
boolean_mask = (df['class'] == 'p') & (df[df_columns[20]] == 'n')

In [37]:
joint_probability_denom = df[boolean_mask].shape[0]/df.shape[0]

In [38]:
joint_probability_denom

0.027572624322993598

# Therefore, the final probability: 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n)
\end{equation}

# is given by 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n) = \frac{joint probability num}{joint probability denom}
\end{equation}

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n) = \frac{0.021664204825209258}{0.027572624322993598}
\end{equation}

In [39]:
conditional_probability = joint_probability_num/joint_probability_denom

In [40]:
conditional_probability

0.7857142857142858

# Therefore, 

\begin{equation}
P(gill-spacing=c \cap stalk-surface-below-ring=s | class = poissonous \cap spore-print-color=n) = 0.7857142857142858
\end{equation}

# The functions on DataFrame which we are going to talk about: 

# Allign

# Append

# Apply 

# Applymap 

# Combine

# corr

# corrwith

# cov

# describe

# diff 

# drop 

# dropna

# drop_duplicates

# fillna

# filter

# groupby

# head

# hist

# index

# interpolate

# join 

# max 

# mean 

# median 

# merge 

# min 

# mode

# multiply

# pipe 

# pivot 

# pivot_table

# quantile

# replace

# rolling 

# shape 

# shift

# std 

# where 

# tail

# to_csv

# to_numpy

In [41]:
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [42]:
df.tail()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[5 rows x 23 columns]

In [55]:
import numpy as np

In [56]:
new_df1 = pd.DataFrame(data={'column 1':np.random.uniform(size=(10,)), 'column 2': np.random.uniform(size=(10,))})

In [82]:
new_df1

column 1  column 2
0  0.156910  0.763058
1  0.942353  0.365543
2  0.802790  0.119012
3  0.796858  0.485757
4  0.794452  0.113903
5  0.549798  0.003263
6  0.986697  0.672622
7  0.275643  0.408995
8  0.997607  0.736867
9  0.088989  0.250255

In [83]:
new_df2 = pd.DataFrame(data={'column 3':np.random.uniform(size=(5,)), 'column 4':np.random.uniform(size=(5,))})

In [84]:
new_df2

column 3  column 4
0  0.121634  0.279942
1  0.682977  0.330752
2  0.173508  0.418258
3  0.515562  0.297752
4  0.910800  0.275377

In [85]:
new_df1.align(other=new_df2,axis=0)

(   column 1  column 2
 0  0.156910  0.763058
 1  0.942353  0.365543
 2  0.802790  0.119012
 3  0.796858  0.485757
 4  0.794452  0.113903
 5  0.549798  0.003263
 6  0.986697  0.672622
 7  0.275643  0.408995
 8  0.997607  0.736867
 9  0.088989  0.250255,    column 3  column 4
 0  0.121634  0.279942
 1  0.682977  0.330752
 2  0.173508  0.418258
 3  0.515562  0.297752
 4  0.910800  0.275377
 5       NaN       NaN
 6       NaN       NaN
 7       NaN       NaN
 8       NaN       NaN
 9       NaN       NaN)

In [86]:
another_df = new_df1.append(other=new_df2)

In [87]:
another_df

column 1  column 2  column 3  column 4
0  0.156910  0.763058       NaN       NaN
1  0.942353  0.365543       NaN       NaN
2  0.802790  0.119012       NaN       NaN
3  0.796858  0.485757       NaN       NaN
4  0.794452  0.113903       NaN       NaN
5  0.549798  0.003263       NaN       NaN
6  0.986697  0.672622       NaN       NaN
7  0.275643  0.408995       NaN       NaN
8  0.997607  0.736867       NaN       NaN
9  0.088989  0.250255       NaN       NaN
0       NaN       NaN  0.121634  0.279942
1       NaN       NaN  0.682977  0.330752
2       NaN       NaN  0.173508  0.418258
3       NaN       NaN  0.515562  0.297752
4       NaN       NaN  0.910800  0.275377

In [69]:
D =dict()

In [70]:
D['key'] = 'value'

In [71]:
D

{'key': 'value'}

In [72]:
D['another key'] = 'another value'

In [73]:
D

{'key': 'value', 'another key': 'another value'}

In [90]:
new_df1

column 1  column 2
0  0.156910  0.763058
1  0.942353  0.365543
2  0.802790  0.119012
3  0.796858  0.485757
4  0.794452  0.113903
5  0.549798  0.003263
6  0.986697  0.672622
7  0.275643  0.408995
8  0.997607  0.736867
9  0.088989  0.250255

In [91]:
new_df2

column 3  column 4
0  0.121634  0.279942
1  0.682977  0.330752
2  0.173508  0.418258
3  0.515562  0.297752
4  0.910800  0.275377

In [96]:
first_df = pd.DataFrame(data=np.random.uniform(size=(10,)),columns=['column 1'])

In [97]:
first_df

column 1
0  0.359708
1  0.395457
2  0.367056
3  0.086296
4  0.263198
5  0.766497
6  0.528160
7  0.045037
8  0.757731
9  0.927233

In [98]:
first_df['column 2'] = np.random.uniform(size=(10,))

In [99]:
first_df

column 1  column 2
0  0.359708  0.937558
1  0.395457  0.073726
2  0.367056  0.502369
3  0.086296  0.629705
4  0.263198  0.665207
5  0.766497  0.090679
6  0.528160  0.809006
7  0.045037  0.623854
8  0.757731  0.951079
9  0.927233  0.654354

In [100]:
second_df = pd.DataFrame(data=first_df['column 2'])

In [102]:
second_df['column 3'] = np.random.uniform(size=(10,))

In [103]:
second_df

column 2  column 3
0  0.937558  0.385305
1  0.073726  0.307700
2  0.502369  0.197729
3  0.629705  0.693324
4  0.665207  0.412791
5  0.090679  0.783989
6  0.809006  0.113672
7  0.623854  0.706536
8  0.951079  0.931776
9  0.654354  0.095647

In [104]:
third_df = first_df.append(other=second_df)

In [105]:
third_df

column 1  column 2  column 3
0  0.359708  0.937558       NaN
1  0.395457  0.073726       NaN
2  0.367056  0.502369       NaN
3  0.086296  0.629705       NaN
4  0.263198  0.665207       NaN
5  0.766497  0.090679       NaN
6  0.528160  0.809006       NaN
7  0.045037  0.623854       NaN
8  0.757731  0.951079       NaN
9  0.927233  0.654354       NaN
0       NaN  0.937558  0.385305
1       NaN  0.073726  0.307700
2       NaN  0.502369  0.197729
3       NaN  0.629705  0.693324
4       NaN  0.665207  0.412791
5       NaN  0.090679  0.783989
6       NaN  0.809006  0.113672
7       NaN  0.623854  0.706536
8       NaN  0.951079  0.931776
9       NaN  0.654354  0.095647

# Till now, we have undertstood the functionality behind align as well as append. 

# Now, we are going to understand the functionality behind apply. 

In [107]:
def f(x):
    
    y = 2*x**2 + 5*x + 3
    
    return y

In [108]:
modified_df = first_df.apply(func=f)

In [109]:
first_df

column 1  column 2
0  0.359708  0.937558
1  0.395457  0.073726
2  0.367056  0.502369
3  0.086296  0.629705
4  0.263198  0.665207
5  0.766497  0.090679
6  0.528160  0.809006
7  0.045037  0.623854
8  0.757731  0.951079
9  0.927233  0.654354

In [110]:
modified_df

column 1  column 2
0  5.057320  9.445820
1  5.290061  3.379500
2  5.104739  6.016597
3  3.446375  6.941582
4  4.454538  7.211034
5  8.007522  3.469839
6  6.198703  8.354008
7  3.229244  6.897658
8  7.936970  9.564497
9  9.355691  7.128131